cranelift/codegen/src/data_value.rs

```diff
@@ -0,0 +1,195 @@
+//! This module gives users to instantiate values that Cranelift understands. These values are used,
+//! for example, during interpretation and for wrapping immediates.
+use crate::ir::immediates::{Ieee32, Ieee64, Imm64, Offset32};
+use crate::ir::{types, ConstantData, Type};
+use core::convert::TryInto;
+use core::fmt::{self, Display, Formatter};
+use thiserror::Error;
+
+/// Represent a data value. Where [Value] is an SSA reference, [DataValue] is the type + value
+/// that would be referred to by a [Value].
+///
+/// [Value]: crate::ir::Value
+#[allow(missing_docs)]
+#[derive(Clone, Debug, PartialEq)]
+pub enum DataValue {
+    B(bool),
+    I8(i8),
+    I16(i16),
+    I32(i32),
+    I64(i64),
+    F32(f32),
+    F64(f64),
```

Comment:
What do you think of https://doc.rust-lang.org/std/primitive.f32.html#method.from_bits? Looks to me like `f32::from_bits` (which is how we get from `Ieee32` for this value) is not going to change NaN bits at all.

Reply:
The `to_bits` method can change it. x86 processors normalize it when moving out of an SSE register I think.

Reply:
I don't know... Rust devs think the issue is "overblown": https://doc.rust-lang.org/src/core/num/f32.rs.html#700-702.

Reply:
That is a different issue. What I am talking about here is that for example when the user intended an sNaN it would quietly get turned into a regular NaN.

Reply:
Zulip?

Reply:
To add my 2¢ here: I don't think there's much cost to using `Ieee32`/`Ieee64` (just some bitcasts), it has accessors to give us float values when we need them, it guarantees bit-for-bit accuracy otherwise as we manipulate code, and we already use it elsewhere for constants. I'd prefer the whole lowering path, from CLIF to the final point that we emit constants into the `MachBuffer`, to avoid potential machine-specific issues with `f32`/`f64`. Possiblyl @sunfishcode has thoughts (I know you've talked about NaNs and determinism before)?

Reply:
I'm not opposed to using `Ieee32/Ieee64` here but I do want to sort out whether there are real issues that they avoid or not; if not, then I think Rust's `f32/f64` are a bit simpler and perhaps we should be using those other places, e.g., for CLIF parsing.

Reply:
Here is a summary of what @bjorn3 and I discussed on Zulip: because ARM machines might have [default NaN mode](https://developer.arm.com/documentation/ddi0344/i/neon-and-vfp-programmers-model/modes-of-operation/default-nan-mode?lang=en) set during compilation but not during runtime, the `f32`, if operated on (@bjorn3 thinks even possibly by moving to a register), could lose its signalling bit. This makes me lean toward switching to use `Ieee32` but I would still like to hear @sunfishcode's opinion.

I'm still a bit mystified on why Rust core thinks this will not be an issue in their case: https://github.com/rust-lang/rust/commit/a2cdeb58f680b87b5bdb6a17cba857ac51307c8f#diff-14f327fd5e3b91eb2e9af0690596d156R284.

Reply:
I think what was thought to be a safety issue is sNaN causing a floating point trap as opposed to return sNaN or qNaN bits or even LLVM returning poison.

Reply:
I favor `Ieee32`/`Ieee64` here too.

It's my understanding that Rust as a language is still figuring out its strategy for NaN bits: https://github.com/rust-lang/unsafe-code-guidelines/issues/237 (to be sure, it's a hard problem). Until that's resolved, I'd prefer not to skate near the edge :-).